### ISSUE TITLE
Q2RTX 1.6.0 client crashes with apparent null pointer exception.

### ISSUE THREAD

**Issue by QwazyWabbitWOS** [NVIDIA/Q2RTX#198]
Client crashes in `find_or_load_image` function after initialization and connection to servers.
[Q2RTX_CrashReport26.txt](https://github.com/NVIDIA/Q2RTX/files/8149451/Q2RTX_CrashReport26.txt)

**Comment by res2k** [comment]
What did you do before the crash? Can you provide instructions on how to trigger it?

**Comment by QwazyWabbitWOS** [comment]
I used an ancient program called All Seeing Eye to launch Q2RTX. Yahoo bought the project from the developer and re-brandeded it and then dropped support for it when the developer's soul didn't come with it. It is a server browser monitor like QTracker that formulates a command line aiming the Q2RTX client at the selected server. It's been working just fine with all clients: r1q2, q2pro, older versions of Q2RTX but suddenly started malfunctioning with 1.6.0. Launching the client directly from Windows Explorer or the Start dialog, then using Q2RTX server scanner only crashes intermittently but it still crashes. It looks like the 'file' argument to the function is null.

**Comment by res2k** [comment]
> using Q2RTX server scanner only crashes intermittently but it still crashes

You mean "Browse q2servers.com"? 
Does it crash just opening the list? That'd be something to go on...

Otherwise, any way you can narrow it down further (e.g. some specific server(s), map, games...) would be helpful.

**Comment by QwazyWabbitWOS** [comment]
The client is fine until it's in the connection process with the server and is negotiating the file cache. If you can suggest a way to attach VS2019 debugger to the process and break it at the exception to trace the stack that would be helpful, otherwise I don't know what else to tell you about why the argument to the function is bad.

**Comment by QwazyWabbitWOS** [comment]
I put an assertion statement in `find_or_load_image` and dropped the new client executable into the game root and as you might expect, it's launching and playing just fine. I've been seeing this issue for a couple of weeks now, ever since the 1.6.0 release. I don't know if it's a code or content bug.

**Comment by QwazyWabbitWOS** [comment]
@res2k Frank, I think I have a reproducible scenario. 
It looks like q2proded is locking local content files.
It requires this setup:
Start a q2pro dedicated server on localhost. It doesn't seem to matter what mod.
Start a q2rtx client from within the same quake2 root and drop console. Command `connect localhost` or 127.0.0.1 the connection proceeds until the exception dialog pops.

Running a dedicated local r1q2 server instead of q2pro does not induce this fault.

**Comment by res2k** [comment]
Hmm, no "luck" yet...
- Downloaded binaries from https://skuller.net/q2pro/
- Dropped q2proded.exe into my Steam Q2RTX directory
- Started server there
- Started release Q2RTX, did `connect localhost` a couple of times ... no crash

Perhaps collecting some additional info with Visual Studio might not be a bad thing.
Did you build Q2RTX from source? If so, you could paste the callstack from VS? Line numbers tend to be very useful ;)
Also good to know would be the name of the problematic file/material, that may give a hint what's going on.
Or anything else that looks suspicious, for that matter.

**Comment by QwazyWabbitWOS** [comment]
I'll see what I can do and get back to you. That crash report was from the 1.6.0 release binary and pdb co-located in the same directory with the q2proded server. I can't make it fault by running the RTX client in the debugger. I don't think I've ever seen it happen there. As for stack trace, the one in the crash reports is all I have at the moment.

**Comment by QwazyWabbitWOS** [comment]
Here we go...

[debug_stack_data.txt](https://github.com/NVIDIA/Q2RTX/files/8153879/debug_stack_data.txt)

**Comment by res2k** [comment]
Hm, this is weird...
The code around the crashing line is:
```c
	int override_textures = !!r_override_textures->integer;
	if (!vid_rtx->integer && (type != IT_PIC) && !gl_use_hd_assets->integer)
		override_textures = 0;
    if (flags & IF_EXACT)
        override_textures = 0;
```
So my suspicion would be that the segfault happens either accessing `vid_rtx` or `gl_use_hd_assets` are `NULL`.
Could you check that the next time?

Also, maybe it's a good time to talk about what settings you're using - RTX vs GL and HD assets on vs off may be relevant.
Other stuff possibly as well... did you change any settings to values that might be considered unusual?

**Comment by res2k** [comment]
> Could you check that the next time?

Actually, if non-`NULL`, please also check their members - does everything look sensible or does something seem off...

**Comment by QwazyWabbitWOS** [comment]
`vid_rtx->integer` is 0, I'm using openGL at the moment. 
I've been using opengl for quite some time now so that's not a recent change here.
I have the debugger stopped and open at the trap so I can give you any values you need right now.
`gl_use_hd_assets` is NULL.
 `find_or_load_image` has been called with &image = NULL.
 `override_textures` is 1.

**Comment by res2k** [comment]
`*image_p == NULL` is fine, this will be written if the functions succeeds.
But `gl_use_hd_assets` being `NULL`, that is indeed weird.

It should certainly be initialized in in GL mode, see src/refresh/gl/main.c:732.
One idea would be to add some console outputs to that function, see if it gets called and/or `gl_use_hd_assets` gets assigned a proper (pointer) value.

**Comment by QwazyWabbitWOS** [comment]
For some reason, `GL_Register` is not being called when q2rtx is invoked from Explorer (double-click)
or with a command line. (e.g. `r2rtx connect localhost`) or via game launchers like ASE or QTracker.
All the `gl_*` cvars in that function are null after the exception is thrown.
Race condition?

**Comment by res2k** [comment]
Hm, race condition seems unlikely to me... mainly because Q2RTX is, for all intents and purposes, single-threaded.

`GL_Register()` not being called is _extremely_ weird. It's typically called as part of the initialization.
See here for an example stack:
```
q2rtx.exe!GL_Register() Line 709 (d:\sources\Q2RTX\src\refresh\gl\main.c:709)
q2rtx.exe!R_Init_GL(bool total) Line 860 (d:\sources\Q2RTX\src\refresh\gl\main.c:860)
q2rtx.exe!CL_InitRefresh() Line 348 (d:\sources\Q2RTX\src\client\refresh.c:348)
q2rtx.exe!CL_Init() Line 3420 (d:\sources\Q2RTX\src\client\main.c:3420)
q2rtx.exe!Qcommon_Init(int argc, char * * argv) Line 1024 (d:\sources\Q2RTX\src\common\common.c:1024)
q2rtx.exe!Sys_Main(int argc, char * * argv) Line 1434 (d:\sources\Q2RTX\src\windows\system.c:1434)
q2rtx.exe!WinMain(HINSTANCE__ * hInstance, HINSTANCE__ * hPrevInstance, char * lpCmdLine, int nCmdShow) Line 1508 (d:\sources\Q2RTX\src\windows\system.c:1508)
q2rtx.exe!invoke_main() Line 107 (d:\a01\_work\12\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:107)
q2rtx.exe!__scrt_common_main_seh() Line 288 (d:\a01\_work\12\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
q2rtx.exe!__scrt_common_main() Line 331 (d:\a01\_work\12\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:331)
q2rtx.exe!WinMainCRTStartup(void * __formal) Line 17 (d:\a01\_work\12\s\src\vctools\crt\vcstartup\src\startup\exe_winmain.cpp:17)
kernel32.dll!BaseThreadInitThunk() (Unknown Source:0)
ntdll.dll!RtlUserThreadStart() (Unknown Source:0)
```

There's very little that executes before that, and I don't see a reason why `GL_Register()` would be skipped - except maybe   `vid_rtx` being `!= 0`, but that means RTX rendering should be used, not GL!

**Comment by QwazyWabbitWOS** [comment]
Yeah, I'm used to dealing with CNC & PLC code where I/O state affects branching. Sometimes I forget Q2 is single thread.
I'll set some debug prints and try to find out why this is happening.

**Comment by QwazyWabbitWOS** [comment]
@res2k I have found the root cause.
When launching from command line or from a game launcher like ASE or QTracker, the Q2RTX client loads the q2config.cfg from baseq2 folder before any q2config.cfg it might find in the mod folder for the targeted server. If that q2config.cfg has `seta vid_rtx 1` in it and if the user has `set vid_rtx 0` in their autoexec.cfg there is a double initialization. The first sets up RTX mode and the second sets up OpenGL mode but because the initialization has already occurred, the `CL_InitRefresh` function doesn't call `R_RegisterFunctionsGL` and can never call `GL_Register`, leaving all the GL cvars uninitialized.

IMO, it is a BadPractice:tm: to fail to initialize cvars. The whole point of them is that they are never NULL pointers or hold garbage values.

**Comment by res2k** [comment]
Hmm, curious...
But there might be another piece of the puzzle missing here... I tried to reproduce it by having different `vid_rtx` values in `q2config.cfg` vs `autoexec.cfg`, but so far I couldn't trigger the crash. I'll keep trying, but any additional details you can provide may be helpful.

**Comment by QwazyWabbitWOS** [comment]
The only thing I can think of is that I have copied the retail q2rtx client and the content files to my quake2 root rather than using the ones in the "installed" Games\Quake2RTX folder. I didn't want to have duplicate mod content everywhere on C drive and I already owned the Quake2 retail release for 20+ years. The other difference is that the dedicated server is on the same tree as the client so perhaps file locking of q2config might be involved and the client can't write to it?

I can duplicate the crash by simply setting 1 on `vid_rtx` in the baseq2\q2config.cfg and remedy it by setting 0. No other changes at all. Using the Video menu to change modes is not a problem.

**Comment by QwazyWabbitWOS** [comment]
When I run the client in the debugger I get different behavior than from same client in the quake2 folder. I'm in the same boat as you with inability to reproduce the fault in the debugger.

When the client is failing for me there is a white rectangle inside a black fullscreen app, then a brief view of the console, followed another white on black then a crash with the null ptr. Correcting the config with `vid_rtx 0`, the retail client works correctly, the white on black period is shorter and the console proceeds normally.

The debugger instance rewrites its copy of the config to `set vid_rtx 0` at startup, it doesn't wait until shutdown. The original  quake2 wrote it's config.cfg file out at client shutdown in order to capture changes made during gameplay so that on the next startup they could be reinstated unless overridden in the autoexec.cfg. As you can see from the log, the client restarts the VFS and this forces a re-write of the config. You need to be aware of this and to reset the initial state before each launch of the client. 

Making the q2config.cfg file read-only confirms this, producing a message in the console.log at the startup that it couldn't write to the file during the start-up & connect:
...
Requesting challenge... 1
====== Quake II RTX initialized ======

Requesting connection... 2
[HTTP] Download server at http://q2http.tastyspleen.net/
Connected to 127.0.0.1:27910 (protocol 36).
**Couldn't open q2config.cfg for writing: Permission denied**
----- FS_Restart -----
...

I assert that the startup code needs rework. All cvars should be unconditionally initialized at startup irrespective of the renderer in use. Fixing just this might actually resolve the whole issue.

[console.log](https://github.com/NVIDIA/Q2RTX/files/8175293/console.log)

**Comment by res2k** [comment]
Alright, I was finally able to replicate the issue, and I think I fixed it in PR#199.
Turns out the culprit was an incorrect usage of the `vid_rtx` value - it was assumed this would reflect the currently active renderer, but it doesn't necessarily do so.
In your use case, with `vid_rtx` in changed in different config files, it was set to 0 before the renderer was actually changed; so RTX was active, but some code thought it GL would be active and behaving as such (leading to the crash).
